### PR TITLE
Allow data outputs in PS collaterals

### PR DESCRIPTION
**Pull request by UdjinM6**
This should allow to
- reduce utxo spam caused by 1x collaterals a bit
- use otherwise non-usable small inputs (in the range of 1-2x) as PS collaterals, which should reduce utxo set even further

### CODE REVIEW

File: src/privatesend.cpp [hunk 1]
```diff
@@ -304,10 +304,9 @@ bool CPrivateSend::IsCollateralValid(const CTransaction& txCollateral)
 
 bool CPrivateSend::IsCollateralAmount(CAmount nInputAmount)
 {
-    // collateral inputs should always be a 2x..4x of mixing collateral
-    return  nInputAmount >  GetCollateralAmount() &&
-            nInputAmount <= GetMaxCollateralAmount() &&
-            nInputAmount %  GetCollateralAmount() == 0;
+    // collateral input should be smth between 1x and 2x OR exactly Nx
+    return (nInputAmount > GetCollateralAmount() && nInputAmount < GetCollateralAmount()*2) ||
```
Comment (codablock) [inline src/privatesend.cpp]: Spaces around `*` missing.

File: src/privatesend.cpp [hunk 2]
```diff
@@ -305,8 +306,9 @@ bool CPrivateSend::IsCollateralValid(const CTransaction& txCollateral)
 bool CPrivateSend::IsCollateralAmount(CAmount nInputAmount)
 {
     // collateral input should be smth between 1x and 2x OR exactly Nx
-    return (nInputAmount > GetCollateralAmount() && nInputAmount < GetCollateralAmount()*2) ||
-           (nInputAmount >= GetCollateralAmount() && nInputAmount <= GetMaxCollateralAmount() && nInputAmount % GetCollateralAmount() == 0);
+    int nMin = mnpayments.GetMinMasternodePaymentsProto() > 70208 ? 1 : 2;
```
Comment (codablock) [inline src/privatesend.cpp]: This won't disable the change in requirement for exact Nx collaterals. In the old version it was not allowed to have something between 1x and 2x as I understand it.

Comment (codablock) [reply src/privatesend.cpp]: Maybe it's better to do a simple `if (mnpayments.GetMinMasternodePaymentsProto() > 70208) {} else {}` here? Maybe better readable and at the same time more obvious later when we clean up the code.

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -3323,18 +3323,22 @@ bool CWallet::CreateCollateralTransaction(CMutableTransaction& txCollateral, std
...
     txCollateral.vin.push_back(txdsinCollateral);
 
     //pay collateral charge in fees
-    CTxOut txout = CTxOut(nValue - CPrivateSend::GetCollateralAmount(), scriptChange);
-    txCollateral.vout.push_back(txout);
+    if (nValue >= CPrivateSend::GetCollateralAmount()*2) {
+        // make our change address
+        CScript scriptChange;
+        CPubKey vchPubKey;
+        assert(reservekey.GetReservedKey(vchPubKey, true)); // should never fail, as we just unlocked
```
Comment (codablock) [inline src/wallet/wallet.cpp]: Not introduced by this PR, but in Bitcoin they removed all uses of `assert(doImportantLogic())` and replaced it with `if(!doImportantLogic()) assert(false);` or `bool success = doImportantLogic(); assert(success);`
I assume they started doing this to allow release builds which omit all asserts.

File: src/wallet/wallet.cpp [hunk 4]
```diff
@@ -3323,18 +3323,22 @@ bool CWallet::CreateCollateralTransaction(CMutableTransaction& txCollateral, std
...
+        scriptChange = GetScriptForDestination(vchPubKey.GetID());
+        reservekey.KeepKey();
+        // return change
+        txCollateral.vout.push_back(CTxOut(nValue - CPrivateSend::GetCollateralAmount(), scriptChange));
+    } else { // nValue < CPrivateSend::GetCollateralAmount()*2
+        // create dummy data output only and pay everything as a fee
+        txCollateral.vout.push_back(CTxOut(0, CScript() << OP_RETURN));
```
Comment (codablock) [inline src/wallet/wallet.cpp]: Does this require the protobump patch as well?

File: src/wallet/wallet.cpp [hunk 5]
```diff
@@ -3323,18 +3323,22 @@ bool CWallet::CreateCollateralTransaction(CMutableTransaction& txCollateral, std
...
-    assert(reservekey.GetReservedKey(vchPubKey, true)); // should never fail, as we just unlocked
-    scriptChange = GetScriptForDestination(vchPubKey.GetID());
-    reservekey.KeepKey();
-
     txCollateral.vin.push_back(txdsinCollateral);
 
     //pay collateral charge in fees
-    CTxOut txout = CTxOut(nValue - CPrivateSend::GetCollateralAmount(), scriptChange);
-    txCollateral.vout.push_back(txout);
+    if (nValue >= CPrivateSend::GetCollateralAmount()*2) {
```
Comment (codablock) [inline src/wallet/wallet.cpp]: Missing space around `*`

**Comment by UdjinM6** [comment]
Thanks! Addressed review comments in 5f9367a.

**Comment by UdjinM6** [comment]
That's a good question actually. Nx is a legacy thing which used to be used to identify collateral inputs when there were no denoms (you could mix any amounts in early version of PS). This was an optimisation to avoid looking for rounds for such inputs and also to easily spot them in CoinControl. It's not really needed now because rounds calculation algo was heavily reworked since then and you probably can live without seeing collateral in CoinControl... but no one was curious enough to spot this thing yet I guess ;) It's ok to use smth between 1x and 2x with data output even with Nx thing because it doesn't create new output (it pays everything as a fee), so there is no overhead or privacy concerns. But thinking more about it I guess using single inputs up to 4x (paying 1x as a fee and everything else as a change) should also be fine... 🤔

**Comment by UdjinM6** [comment]
Actually, it seems that non-Nx inputs which are greater than `1x + dust` should already be accepted by the server side (MNs), it's just that the client side won't create those. This could probably be yet another good tweak to consume small inputs as collaterals and lower utxo bloat even further but let's leave this for another PR :)

**Pull request merged**